tests/src/junit/java/net/neoforged/neoforge/unittest/transfer/ItemTests.java

```diff
@@ -0,0 +1,261 @@
+        }
+
+        @Override
+        public boolean canTakeItemThroughFace(int slot, ItemStack stack, Direction dir) {
+            return dir != Direction.UP;
+        }
+    }
+
+    /**
+     * Test insertion when {@link Container#getMaxStackSize()} is the bottleneck.
+     */
+    @Test
+    public void testLimitedStackCountContainer() {
+        ItemResource diamond = ItemResource.of(Items.DIAMOND);
+        LimitedStackCountContainer container = new LimitedStackCountContainer(diamond.toStack(), diamond.toStack(), diamond.toStack());
+        var wrapper = VanillaContainerWrapper.of(container);
+
+        // Should only be able to insert 2 diamonds per stack * 3 stacks = 6 diamonds.
+        try (Transaction transaction = Transaction.open(null)) {
+            if (wrapper.insert(diamond, 1000, transaction) != 6) {
+                throw new AssertionError("Only 6 diamonds should have been inserted.");
+            }
+
+            checkComparatorOutput(container);
+        }
+    }
+
+    /**
+     * Test insertion when {@link Item#getMaxStackSize} is the bottleneck.
+     */
+    @Test
+    public void testLimitedStackCountItem() {
+        ItemResource diamondPickaxe = ItemResource.of(Items.DIAMOND_PICKAXE);
+        LimitedStackCountContainer container = new LimitedStackCountContainer(5);
+        var wrapper = VanillaContainerWrapper.of(container);
+
+        // Should only be able to insert 5 pickaxes, as the item limits stack counts to 1.
+        try (Transaction transaction = Transaction.open(null)) {
+            if (wrapper.insert(diamondPickaxe, 1000, transaction) != 5) {
+                throw new AssertionError("Only 5 pickaxes should have been inserted.");
+            }
+
+            checkComparatorOutput(container);
+        }
+    }
+
+    private static class LimitedStackCountContainer extends SimpleContainer {
+        LimitedStackCountContainer(int size) {
+            super(size);
+        }
+
+        LimitedStackCountContainer(ItemStack... stacks) {
+            super(stacks);
+        }
+
+        @Override
+        public int getMaxStackSize() {
+            return 3;
+        }
+    }
+
+    private static void checkComparatorOutput(Container container) {
+        var wrapper = VanillaContainerWrapper.of(container);
+
+        int vanillaOutput = AbstractContainerMenu.getRedstoneSignalFromContainer(container);
+        int transferApiOutput = ResourceHandlerUtil.getRedstoneSignalFromResourceHandler(wrapper);
+
+        if (vanillaOutput != transferApiOutput) {
+            String error = String.format(
+                    "Vanilla and Transfer API comparator outputs should have been identical. Vanilla: %d. Transfer API: %d.",
+                    vanillaOutput,
+                    transferApiOutput);
+            throw new AssertionError(error);
+        }
+    }
+
+    /**
+     * Ensure that SimpleContainer only calls setChanged at the end of a successful transaction.
```

Comment:
Do we want to be testing it only gets called for the *root* success? And not nested transactions, or does those kinds of tests happen elsewhere?

Reply:
Since it's based on an `onRootCommit` override I think it's already tested through other means.